### PR TITLE
[Merge] ExecutionEngineService refactor

### PR DESCRIPTION
**Pull request by Nashatyrev**
## PR Description

- Extract `ExecutionEngineService` channel interface `ExecutionEngineChannel` which is the only left in the `spec` module 
- Move all other implementation classes to the `powchain` service module 
- Instead of injecting `ExecutionEngineChannel` into the spec (which is not great) pass the `ExecutionEngineChannel` to various components upon creation and then pass it as an argument to the spec functions: remove the `ExecutionEngineChannel` filed from `ExecutionPayloadUtil` and `MergeTransitionHelpers`
- Move all the logic from `BeaconChainController.initExecutionEngineService()` to the `ExecutionEngineService`

Still problems: 
- [x] `ExecutionEngineChannel` is now the part of the `spec` (should it really be). Need a proper way to pass the implementation to the spec structures 
- [x] `BlockProcessorMerge.forProcessExecutionPayloadReferenceTest()` should be moved to the `ref-tests` module
- [x] Move all the logic from `BeaconChainController.initExecutionEngineService()` to the `ExecutionEngineService`

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: services/powchain/src/main/java/tech/pegasys/teku/services/powchain/execution/ExecutionEngineService.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  private static <K> K unwrapResponseOrThrow(Response<K> response) {
+    checkArgument(
+        response.getPayload() != null, "Invalid remote response: %s", response.getReason());
+    return response.getPayload();
+  }
+
+  @FormatMethod
+  private static void printConsole(String formatString, Object... args) {
+    LOG.info(ColorConsolePrinter.print(String.format(formatString, args), Color.CYAN));
+  }
```
Comment (ajsutton) [inline services/powchain/src/main/java/tech/pegasys/teku/services/powchain/execution/ExecutionEngineService.java]: This seems useful for debugging but likely becomes trace level in actual production code right?

**Review by ajsutton — COMMENTED** [review]
Yep this looks like the kind of thing I had in mind. Injecting the `ExecutionEngineChannel` publisher into the Spec is going to be a bit awful unfortunately - I think it either has to pass through a whole bunch of static `create` methods as an extra parameter or move to more of a builder pattern for Spec, though I'm not sure that simplifies things.

**Comment by ajsutton** [comment]
So thinking about this I don't think it will be possible to pass the execution engine publisher into the spec when it's created. It's created much to early in the startup process and from far too many places. Plus it will be really common to want to mock it which is hard to do when it's wrapped up in the spec. 

So I suspect we need to move to passing in the execution engine when calling the methods that will need it rather than having it be contained within the spec.

File: services/powchain/src/main/java/tech/pegasys/teku/services/powchain/execution/ExecutionEngineService.java (hunk 1 above)
Comment (Nashatyrev) [reply services/powchain/src/main/java/tech/pegasys/teku/services/powchain/execution/ExecutionEngineService.java]: Yep, sure!

Comment (Nashatyrev) [reply services/powchain/src/main/java/tech/pegasys/teku/services/powchain/execution/ExecutionEngineService.java]: Just tried to preserve the original code. Would be probably useful for interop

File: eth-reference-tests/src/referenceTest/java/tech/pegasys/teku/reference/ManualReferenceTestRunner.java [hunk 2]
```diff
@@ -59,6 +59,7 @@ void shouldRunReferenceTest(final String name, final TestDefinition testDefiniti
             testDefinition ->
                 SPEC.isBlank() || testDefinition.getConfigName().equalsIgnoreCase(SPEC))
         .filter(testDefinition -> testDefinition.getTestType().startsWith(TEST_TYPE))
+        .filter(testDefinition -> testDefinition.getFork().equals("merge"))
```
Comment (tbenr) [inline eth-reference-tests/src/referenceTest/java/tech/pegasys/teku/reference/ManualReferenceTestRunner.java]: is this needed?

File: ethereum/dataproviders/src/main/java/tech/pegasys/teku/dataproviders/generators/ChainStateGenerator.java [hunk 3]
```diff
@@ -20,17 +20,21 @@
...
   private final BeaconState baseState;
+  private final ExecutionEngineChannel executionEngineChannel;
 
   private ChainStateGenerator(
       final Spec spec,
       final List<SignedBeaconBlock> chain,
       final BeaconState baseState,
-      final boolean skipValidation) {
+      final boolean skipValidation,
+      ExecutionEngineChannel executionEngineChannel) {
```
Comment (tbenr) [inline ethereum/dataproviders/src/main/java/tech/pegasys/teku/dataproviders/generators/ChainStateGenerator.java]: shouldn't `final` like the other params? (occurs in other methods too)

File: eth-reference-tests/src/referenceTest/java/tech/pegasys/teku/reference/ManualReferenceTestRunner.java (hunk 2 above)
Comment (Nashatyrev) [reply eth-reference-tests/src/referenceTest/java/tech/pegasys/teku/reference/ManualReferenceTestRunner.java]: Nope, just accidentally committed. Thanks!

**Comment by Nashatyrev** [comment]
> So I suspect we need to move to passing in the execution engine when calling the methods that will need it rather than having it be contained within the spec.

@ajsutton thanks for advice! A lot of components where we need to pass `ExecutionEngineChannel` but I suppose it's ok. I believe we probably would need to do something similar for `DAS` later.

File: services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java [hunk 4]
```diff
@@ -806,67 +806,20 @@ private void createExecutionEngineChannel() {
     Optional<SpecLogic> specLogic = spec.getSpecLogicForMilestone(SpecMilestone.MERGE);
     if (specLogic.isPresent() && beaconConfig.powchainConfig().isEnabled()) {
       executionEngineChannel =
-          ExecutionEngineService.create(beaconConfig.powchainConfig().getEth1Endpoints().get(0));
+          ExecutionEngineChannelImpl.create(
+              beaconConfig.powchainConfig().getEth1Endpoints().get(0));
```
Comment (tbenr) [inline services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java:810]: there will be maybe a future config dedicated to the execution engine endpoint?

**Comment by tbenr** [comment]
LGTM (btw not fully aware of everything)

**Review by tbenr — APPROVED** [review]

File: services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java (hunk 4 above)
Comment (Nashatyrev) [reply services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java:810]: Yep, I think this is more like a stub option now (just copied it from original version). 
@mkalinin do you think it's worth leaving this 'as is' for interop ?

**Review by mkalinin — APPROVED** [review]
Looks good to me 👍

**Pull request merged**